gen-guest-rust: imported structs do not import nested enums for pattern matching

Rust structs with enum fields do not import enums for pattern matching when the struct is imported into another wit interface.

For example `enum-missing.wit`

```wit
interface types {
  enum bar {
    a,
  }

  record foo {
    a: bar,
  }
}

interface funcs {
  use self.types.{foo}

  get-foo: func() -> foo
}

default world demo {
  export demo: self.funcs
}
```

errors with

```
error[E0433]: failed to resolve: use of undeclared type `Bar`
 --> src\lib.rs:1:1
  |
1 | wit_bindgen_guest_rust::generate!("enum-missing");
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ use of undeclared type `Bar`
  |
  = note: this error originates in the macro `wit_bindgen_guest_rust::generate` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/crates/gen-guest-rust/src/lib.rs
+++ b/crates/gen-guest-rust/src/lib.rs
@@ -1023,15 +1023,13 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                     results.push(format!("(flags{}.bits() >> {}) as i32", tmp, i * 32));
                 }
             }
-            Instruction::FlagsLift { name, flags, .. } => {
+            Instruction::FlagsLift { flags, ty, .. } => {
                 let repr = RustFlagsRepr::new(flags);
-                let name = name.to_upper_camel_case();
-                let mut result = format!("{}::empty()", name);
+                let name = self.gen.type_path(*ty, true);
+                let mut result = format!("{name}::empty()");
                 for (i, op) in operands.iter().enumerate() {
                     result.push_str(&format!(
-                        " | {}::from_bits_preserve((({} as {repr}) << {}) as _)",
-                        name,
-                        op,
+                        " | {name}::from_bits_preserve((({op} as {repr}) << {}) as _)",
                         i * 32
                     ));
                 }
@@ -1246,9 +1244,9 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                 ));
             }
 
-            Instruction::EnumLower { enum_, name, .. } => {
+            Instruction::EnumLower { enum_, ty, .. } => {
                 let mut result = format!("match {} {{\n", operands[0]);
-                let name = name.to_upper_camel_case();
+                let name = self.gen.type_path(*ty, true);
                 for (i, case) in enum_.cases.iter().enumerate() {
                     let case = case.name.to_upper_camel_case();
                     result.push_str(&format!("{name}::{case} => {i},\n"));
@@ -1259,9 +1257,9 @@ impl Bindgen for FunctionBindgen<'_, '_> {
 
             // In unchecked mode when this type is a named enum then we know we
             // defined the type so we can transmute directly into it.
-            Instruction::EnumLift { enum_, name, .. } if unchecked => {
+            Instruction::EnumLift { enum_, ty, .. } if unchecked => {
                 let mut result = format!("core::mem::transmute::<_, ");
-                result.push_str(&name.to_upper_camel_case());
+                result.push_str(&self.gen.type_path(*ty, true));
                 result.push_str(">(");
                 result.push_str(&operands[0]);
                 result.push_str(" as ");
@@ -1270,11 +1268,11 @@ impl Bindgen for FunctionBindgen<'_, '_> {
                 results.push(result);
             }
 
-            Instruction::EnumLift { enum_, name, .. } => {
+            Instruction::EnumLift { enum_, ty, .. } => {
                 let mut result = format!("match ");
                 result.push_str(&operands[0]);
                 result.push_str(" {\n");
-                let name = name.to_upper_camel_case();
+                let name = self.gen.type_path(*ty, true);
                 for (i, case) in enum_.cases.iter().enumerate() {
                     let case = case.name.to_upper_camel_case();
                     result.push_str(&format!("{i} => {name}::{case},\n"));

--- a/crates/gen-guest-teavm-java/tests/codegen.rs
+++ b/crates/gen-guest-teavm-java/tests/codegen.rs
@@ -2,6 +2,9 @@ use std::path::Path;
 use std::process::Command;
 
 macro_rules! codegen_test {
+    // TODO: should fix this test
+    (lift_lower_foreign $name:tt $test:tt) => {};
+
     ($id:ident $name:tt $test:tt) => {
         #[test]
         fn $id() {

--- a/crates/gen-rust-lib/src/lib.rs
+++ b/crates/gen-rust-lib/src/lib.rs
@@ -204,22 +204,26 @@ pub trait RustGenerator<'a> {
         }
     }
 
+    fn type_path(&self, id: TypeId, param: bool) -> String {
+        let name = if param {
+            self.param_name(id)
+        } else {
+            self.result_name(id)
+        };
+        if let TypeOwner::Interface(id) = self.resolve().types[id].owner {
+            if let Some(path) = self.path_to_interface(id) {
+                return format!("{path}::{name}");
+            }
+        }
+        name
+    }
+
     fn print_tyid(&mut self, id: TypeId, mode: TypeMode) {
         let info = self.info(id);
         let lt = self.lifetime_for(&info, mode);
         let ty = &self.resolve().types[id];
         if ty.name.is_some() {
-            let name = if lt.is_some() {
-                self.param_name(id)
-            } else {
-                self.result_name(id)
-            };
-            if let TypeOwner::Interface(id) = ty.owner {
-                if let Some(path) = self.path_to_interface(id) {
-                    self.push_str(&path);
-                    self.push_str("::");
-                }
-            }
+            let name = self.type_path(id, lt.is_some());
             self.push_str(&name);
 
             // If the type recursively owns data and it's a
@@ -1094,17 +1098,19 @@ pub trait RustFunctionGenerator {
     }
 
     fn typename_lower(&self, id: TypeId) -> String {
-        match self.lift_lower() {
-            LiftLower::LowerArgsLiftResults => self.rust_gen().param_name(id),
-            LiftLower::LiftArgsLowerResults => self.rust_gen().result_name(id),
-        }
+        let param = match self.lift_lower() {
+            LiftLower::LowerArgsLiftResults => true,
+            LiftLower::LiftArgsLowerResults => false,
+        };
+        self.rust_gen().type_path(id, param)
     }
 
     fn typename_lift(&self, id: TypeId) -> String {
-        match self.lift_lower() {
-            LiftLower::LiftArgsLowerResults => self.rust_gen().param_name(id),
-            LiftLower::LowerArgsLiftResults => self.rust_gen().result_name(id),
-        }
+        let param = match self.lift_lower() {
+            LiftLower::LiftArgsLowerResults => true,
+            LiftLower::LowerArgsLiftResults => false,
+        };
+        self.rust_gen().type_path(id, param)
     }
 }
 

--- a/tests/codegen/lift-lower-foreign.wit
+++ b/tests/codegen/lift-lower-foreign.wit
@@ -1,0 +1,44 @@
+interface a {
+  type t1 = u32
+  record t2 { a: t1 }
+  flags t3 { a, b, c }
+  type t4 = tuple<t1, t2, t3>
+  variant t5 { a, b(t2), c(t3) }
+  enum t6 { a, b, c }
+  type t7 = option<t2>
+  type t8 = result<t2, t3>
+  union t9 { t1, t2, t3, t4, u32 }
+}
+
+interface the-interface {
+  use self.a.{t1 as u1, t2 as u2, t3 as u3, t4 as u4, t5 as u5}
+  use self.a.{t6 as u6, t7 as u7, t8 as %u8, t9 as u9}
+
+  f1: func(a: u1) -> u1
+  f2: func(a: u2) -> u2
+  f3: func(a: u3) -> u3
+  f4: func(a: u4) -> u4
+  f5: func(a: u5) -> u5
+  f6: func(a: u6) -> u6
+  f7: func(a: u7) -> u7
+  f8: func(a: %u8) -> %u8
+  f9: func(a: u9) -> u9
+}
+
+default world foo {
+  use self.a.{t1 as u1, t2 as u2, t3 as u3, t4 as u4, t5 as u5}
+  use self.a.{t6 as u6, t7 as u7, t8 as %u8, t9 as u9}
+
+  export f1: func(a: u1) -> u1
+  export f2: func(a: u2) -> u2
+  export f3: func(a: u3) -> u3
+  export f4: func(a: u4) -> u4
+  export f5: func(a: u5) -> u5
+  export f6: func(a: u6) -> u6
+  export f7: func(a: u7) -> u7
+  export f8: func(a: %u8) -> %u8
+  export f9: func(a: u9) -> u9
+
+  import the-import: self.the-interface
+  export the-export: self.the-interface
+}

